src/main/kotlin/org/jetbrains/research/testspark/appstarter/ProjectApplicationUtils.kt

```diff
@@ -0,0 +1,224 @@
+    }
+}
+
+class ConfiguratorContextImpl(
+    private val projectRoot: Path,
+    private val indicator: ProgressIndicatorBase = ProgressIndicatorBase(),
+    private val filesFilter: Predicate<Path> = Predicate { true },
+    private val virtualFilesFilter: Predicate<VirtualFile> = Predicate { true }
+) : ConfiguratorContext {
+    private val logger = LoggerFactory.getLogger(javaClass)
+    override fun getProgressIndicator() = indicator
+    override fun getLogger() = object : CommandLineInspectionProgressReporter {
+        override fun reportError(message: String) {
+            logger.warn("ERROR: $message")
+        }
+
+        override fun reportMessage(minVerboseLevel: Int, message: String) {
+            logger.info("PROGRESS: $message")
+        }
+    }
+
+    override fun getProjectPath() = projectRoot
+    override fun getFilesFilter(): Predicate<Path> = filesFilter
+    override fun getVirtualFilesFilter(): Predicate<VirtualFile> = virtualFilesFilter
+}
+
+object ProjectApplicationUtils {
+
+    private val logger = LoggerFactory.getLogger(javaClass)
+
+    /**
+     * Rewritten from {@link com.intellij.codeInspection.InspectionApplicationBase}.
+     * Implementation which reuse InspectionApplicationBase:
+     *
+     * val app = object : InspectionApplicationBase() {
+     *     fun open(): Project? {
+     *         return this.openProject(projectPath, parentDisposable)
+     *     }
+     * }
+     *
+     * return app.open() ?: throw ProjectApplicationException("Can not open project")
+     */
+    suspend fun openProject(projectPath: Path, parentDisposable: Disposable, projectToClose: Project? = null): Project {
+        ApplicationManager.getApplication().assertIsNonDispatchThread()
+        ApplicationManagerEx.getApplicationEx().isSaveAllowed = false
+
+        LocalFileSystem.getInstance().refreshAndFindFileByPath(
+            FileUtil.toSystemIndependentName(projectPath.toString())
+        ) ?: throw ProjectConfiguratorException("Project directory not found.")
+
+        convertProject(projectPath)
+
+        configureProjectEnvironment(projectPath)
+
+        val project = openOrImport(projectPath, projectToClose, forceOpenInNewFrame = true)
+            ?: throw ProjectConfiguratorException("Can not open or import project from $projectPath.")
+        Disposer.register(parentDisposable) { closeProject(project) }
+
+        waitAllStartupActivitiesPassed(project)
+
+        ApplicationManager.getApplication().invokeAndWait {
+            VirtualFileManager.getInstance().refreshWithoutFileWatcher(false)
+        }
+
+        ApplicationManager.getApplication().invokeAndWait {
+            PatchProjectUtil.patchProject(project)
+        }
+
+        waitForInvokeLaterActivities()
+
+        return project
+    }
+
+    private suspend fun convertProject(projectPath: Path) {
+        val conversionService = ConversionService.getInstance()
```

Comment:
Could you write a small comment on what this method does? What kind of conversion does take place?

Reply:
I can't comment a lot here, because this was just part of the working code that I reused. The original did not contain any comments why this is required.